### PR TITLE
[DIA] Bayesian optimisation

### PR DESCRIPTION
**Pull request by DodoTheDeveloper**
This Pull Request implement bayesian optimisation for hyperparameters.

- [x] Implemented random initialization
- [x] Acquisition (Probability of Improvement)
- [x]  Scoring function (L2Norm)
- [x] Kernel (Square Distance)
- [x] Initializing Model
- [x] Updating Model with each iteration
- [x] Fix bug preventing accurate calculation of means and variances with GP.
- [x] Add index/score map to really so that only the actually calculated losses will be returned.
- [x] Fix issue with Singular Matrix

### CODE REVIEW

File: scripts/staging/bayesian_optimisation/test/bayesianOptimisationSinTest.dml [hunk 1]
```diff
@@ -0,0 +1,47 @@
+# Test
+source("./scripts/staging/bayesian_optimisation/bayesianOptimisation.dml") as bayOpt;
+
+kernel_gaus = function(Matrix[Double] X1, Matrix[Double] X2, Double variance)
+  return (Matrix[Double] result)
+{
+    #TODO: What to do about multiple dimensional matrix?
```
Comment (Shafaq-Siddiqi) [inline scripts/staging/bayesian_optimisation/test/bayesianOptimisationSinTest.dml:7]: Hi,
Just quick feedback, in DML we use two-space indentation.

Comment (DodoTheDeveloper) [reply scripts/staging/bayesian_optimisation/test/bayesianOptimisationSinTest.dml]: Hi and thanks for the Feedback, I applied it in the last commit.

**Comment by Baunsgaard** [comment]
Currently the tests are failing because of missing licenses, these can be simply added at the top of the files.

**Comment by DodoTheDeveloper** [comment]
@Baunsgaard Thanks, I've added them now.

File: src/test/java/org/apache/sysds/test/functions/builtin/BuiltinBayesianOptimisationTest.java [hunk 2]
```diff
@@ -1,3 +1,22 @@
+/*
```
Comment (Shafaq-Siddiqi) [inline src/test/java/org/apache/sysds/test/functions/builtin/BuiltinBayesianOptimisationTest.java:1]: Please reformat the test file. Add some test cases for different parameters of the optimizer such as number of iterations, minimize, and ExecType (CP or SPARK).

**Comment by Shafaq-Siddiqi** [comment]
LGTM,
Hi @DodoTheDeveloper, 
Thank you for your contribution. During the merge, I did some minor fixes like formattings, spelling mistakes and Java tests.

**Comment by DodoTheDeveloper** [comment]
> LGTM,
> Hi @DodoTheDeveloper,
> Thank you for your contribution. During the merge, I did some minor fixes like formattings, spelling mistakes and Java tests.

Thanks, I didn't see your reply in time.

**Pull request closed without merging**